python-sqlite-sqlalchemy/project/examples/example_1/main.py

```diff
@@ -1,104 +1,142 @@
-    """
-    # Target date
-    target_date = datetime.strptime(date_string, "%Y-%m-%d").date()
+def get_books_by_publisher(data, ascending=True):
+    """Returns the books by the associated publisher as a pandas series
 
-    # Iterate through the data and get the data
-    data = []
-    for samples in temperature_data.values():
+    Args:
+        data: The pandas dataframe to get the from
+        ascending: The sorting direction for the returned data.
+        Defaults to True.
 
-        # Iterate through the samples
-        for sample_date, sample in samples.items():
+    Returns:
+        The sorted data as a pandas series
+    """
+    return data.groupby("publisher").size().sort_values(ascending=ascending)
 
-            # Generate a date range for the sample
-            min_date = datetime.strptime(
-                sample_date, "%Y-%m-%d"
-            ).date() + timedelta(days=-3)
-            max_date = datetime.strptime(
-                sample_date, "%Y-%m-%d"
-            ).date() + timedelta(days=3)
-            if min_date <= target_date <= max_date:
-                data.append(float(sample))
 
-    # Get the average temp
-    return sum(data) / len(data)
+def get_authors_by_publisher(data, ascending=True):
+    """Returns the authors by the associated publisher as a Panda series
 
+    Args:
+        data: The pandas dataframe to get the data from
+        ascending: The sorting direction for the returned data.
+        Defaults to True.
 
-def get_average_temp_sorted(direction: str, temperature_data: Dict) -> List:
-    dir = direction.lower()
-    if dir not in ["asc", "desc"]:
-        raise Exception(f"Unknown direction: {direction}")
+    Returns:
+        The sorted data as a pandas series
+    """
+    return (
+        data.assign(name=data.first_name.str.cat(data.last_name, sep=" "))
+        .groupby("publisher")
+        .nunique()
+        .loc[:, "name"]
+        .sort_values(ascending=ascending)
+    )
 
-    results = defaultdict(int)
-    for data in temperature_data.values():
-        for date, value in data.items():
-            results[date] += float(value)
 
-    for date, total in results.items():
-        results[date] = float(total) / float(len(temperature_data.keys()))
+def add_new_book(data, author_name, book_title, publisher_name):
+    """Adds a new book to the system"""
+
+    # Does the book exist?
+    if book_title in data["title"].values:
+        raise ValueError("Book exists", book_title)
+
+    # Does the author exist?
+    first_name, _, last_name = author_name.partition(" ")
+    if not any(
+        data["first_name"].str.contains(first_name)
+        & data["last_name"].str.contains(last_name)
+    ):
+        raise ValueError("No author found", author_name)
+
+    # Does the publisher exist?
+    if publisher_name not in data["publisher"].values:
+        raise ValueError("No publisher found", publisher_name)
```

Comment:
I think we should allow all books that are not actual duplicates? Not only allow authors and publishers that already exists. The following code should be fairly consistent with Example 2:

```suggestion
    # Does the book exist?
    first_name, _, last_name = author_name.partition(" ")
    if any(
        (data.first_name == first_name)
        & (data.last_name == last_name)
        & (data.title == book_title)
        & (data.publisher == publisher_name)
    ):
        raise ValueError(
            "New item exists", author_name, book_title, publisher_name
        )
```

Reply:
Example 1 doesn't update the CSV file, so re-running the program doesn't raise an exception. Example 2 does update the database, so re-running the program (without rebuilding the database) does raise an exception. I think this is reasonable for the `add_new_boo()` function.